pages/api/retrieval.ts

```diff
@@ -0,0 +1,170 @@
+      //A filter expression cannot contain partition key or sort key attributes.
+      //You need to specify those attributes in the key condition expression, not the filter expression.
+      FilterExpression: "FormID = :formID",
+      ProjectionExpression: "FormID,SubmissionID,FormSubmission,Retrieved",
+    };
+    //Get form's responses for formID
+    const formResponses = await db.send(new QueryCommand(getItemsDbParams));
+    //Collecting items submissionIDs for logging and updating items.
+    submissionIDlist = formResponses?.Items?.map((response) => {
+      return response.SubmissionID.S;
+    });
+    if (formResponses && formResponses.Count && formResponses.Count > 0) {
+      //Preparing a query to set records's attribute Retrieved to 1.
+      for (const submissionID of submissionIDlist as string[]) {
+        //Create a new statement object
+        const updateItem = {
+          TableName: "Vault",
+          Key: {
+            SubmissionID: {
+              S: submissionID,
+            },
+            FormID: {
+              S: formID,
+            },
+          },
+          UpdateExpression: "SET Retrieved = :retrieved",
+          ConditionExpression: "SubmissionID = :submission",
+          ExpressionAttributeValues: {
+            ":retrieved": { N: "1" },
+            ":submission": { S: submissionID },
+          },
+          ReturnValues: "NONE",
+        };
+        //Update one item at the time
+        await db.send(new UpdateItemCommand(updateItem));
+      }
+    }
+    //Return form data
+    return res.status(200).json({ responses: formResponses.Items });
+  } catch (error) {
+    //logging submissionIDs
+    logMessage.warn(`Possible data loss for these submissionIDs`);
+    //print submission ids as string
+    logMessage.warn(submissionIDlist ? submissionIDlist.toString() : "Empty submissionID list");
+    logMessage.error(error);
+    return res.status(500).json({ responses: "Error on Server Side" });
+  }
+}
+
+/**
+ * This method will make sure that the incoming request is valid by enforcing a set of
+ * rules.
+ * Request parameters :
+ *  - The maxRecords exists and it ranges between 1 and 10.
+ *  - A formID exists and is valid i.e 78
+ * BearerToken :
+ *  - Must contains an email address
+ *  - It has not expired
+ *  - It exists in our Database
+ *  - The record found in the database includes the email, active, and template_id fields
+ *  - The email on the token record matches the email in the payload.
+ *  - The active flag boolean is true.
+ * @param handler - A method to retrieve a form responses
+ * @returns
+ */
+export const formResponsesReqValidator = (
+  handler: (res: NextApiResponse, formID: string, maxRecords: number) => void
+) => {
+  return async function (req: NextApiRequest, res: NextApiResponse): Promise<unknown> {
+    try {
+      //Default value to 10 if it's undefined
+      const { maxRecords = 10, formID } = req.query;
```

Comment:
Why is this default set as a number, and then on line 123 `as string` is being called? Seems unnecessary.

Reply:
The reasoning behind it is that the parameter `maxRecord` may not exist in query string. In that case, we have to pass the default value as string since `parsInt` accepts string. 
On the other hand, if maxRecords exists but the value is undefined its default value will be 10. And keep in mind that the type of a query parameter is `string | string []`.

Reply:
By the i could have set the default as string as well. but it doesn't remove the fact that we need to cast it as string.

Reply:
The query paramerter is typed as `string | string[]` because if a parameter is repeated in the query they are lumped into an array.  Since in our case we only want formID and maxRecords to show up once in the query we can typecheck using something like:
```
      if (Array.isArray(formID) || Array.isArray(maxRecords))
        return res.status(400).json({ error: "Bad Request" });
```
And since the type of maxRecords has been checked and is not `string[]` it is no longer necessary to typecast.
``` 
const{ maxRecords = "10", formID} = req.query;
//...
// maxRecords will either be a default "10" or be the string value from the req.query
const expectedMaxRecords = parseInt(maxRecords)
```